概要：大多机器学习算法不能处理缺失的特征（即NAN值），所以要先处理特征缺失的问题。

1.首先需要设置matplotlib绘图环境

In [1]:
# To support both python 2 and python 3
from __future__ import division, print_function, unicode_literals

# Common imports
import numpy as np
import os

# to make this notebook's output stable across runs
np.random.seed(42)

# To plot pretty figures
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
mpl.rc('axes', labelsize=14)
mpl.rc('xtick', labelsize=12)
mpl.rc('ytick', labelsize=12)

# Ignore useless warnings (see SciPy issue #5998)
import warnings
warnings.filterwarnings(action="ignore", message="^internal gelsd")

2.导入房价数据。

In [2]:
import pandas as pd

DOWNLOAD_ROOT = "https://raw.githubusercontent.com/ageron/handson-ml/master/"
HOUSING_PATH = os.path.join("datasets", "housing")
HOUSING_URL = DOWNLOAD_ROOT + "datasets/housing/housing.tgz"

def load_housing_data(housing_path=HOUSING_PATH):
    csv_path = os.path.join(housing_path, "housing.csv")
    return pd.read_csv(csv_path)

housing = load_housing_data()
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [ ]:
3.查看是否有缺失的特征。

In [3]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
longitude             20640 non-null float64
latitude              20640 non-null float64
housing_median_age    20640 non-null float64
total_rooms           20640 non-null float64
total_bedrooms        20433 non-null float64
population            20640 non-null float64
households            20640 non-null float64
median_income         20640 non-null float64
median_house_value    20640 non-null float64
ocean_proximity       20640 non-null object
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


我们看到特征total_bedrooms（卧室总数）只有20433个值是非空的，说明该特征有缺失值。

4.处理缺失值一般有三种办法，第一种是将有缺失值的行都删除，第二种是将缺失的特征删除，第三种是用特征的中位值来补足数据。

In [4]:
sample_incomplete_rows = housing[housing.isnull().any(axis=1)].head()
sample_incomplete_rows

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
290    -122.16     37.77                47.0       1256.0             NaN   
341    -122.17     37.75                38.0        992.0             NaN   
538    -122.28     37.78                29.0       5154.0             NaN   
563    -122.24     37.75                45.0        891.0             NaN   
696    -122.10     37.69                41.0        746.0             NaN   

     population  households  median_income  median_house_value ocean_proximity  
290       570.0       218.0         4.3750            161900.0        NEAR BAY  
341       732.0       259.0         1.6196             85100.0        NEAR BAY  
538      3741.0      1273.0         2.5762            173400.0        NEAR BAY  
563       384.0       146.0         4.9489            247100.0        NEAR BAY  
696       387.0       161.0         3.9063            178400.0        NEAR BAY

In [5]:
sample_incomplete_rows.dropna(subset=["total_bedrooms"])    # 第一种

Empty DataFrame
Columns: [longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, median_house_value, ocean_proximity]
Index: []

删除为空的特征后，数据集为空。

In [6]:
sample_incomplete_rows.drop("total_bedrooms", axis=1)       # 第二种

longitude  latitude  housing_median_age  total_rooms  population  \
290    -122.16     37.77                47.0       1256.0       570.0   
341    -122.17     37.75                38.0        992.0       732.0   
538    -122.28     37.78                29.0       5154.0      3741.0   
563    -122.24     37.75                45.0        891.0       384.0   
696    -122.10     37.69                41.0        746.0       387.0   

     households  median_income  median_house_value ocean_proximity  
290       218.0         4.3750            161900.0        NEAR BAY  
341       259.0         1.6196             85100.0        NEAR BAY  
538      1273.0         2.5762            173400.0        NEAR BAY  
563       146.0         4.9489            247100.0        NEAR BAY  
696       161.0         3.9063            178400.0        NEAR BAY

特征total_bedrooms被清除了。

In [7]:
median = housing["total_bedrooms"].median()
sample_incomplete_rows["total_bedrooms"].fillna(median, inplace=True) # 第三种
sample_incomplete_rows

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
290    -122.16     37.77                47.0       1256.0           435.0   
341    -122.17     37.75                38.0        992.0           435.0   
538    -122.28     37.78                29.0       5154.0           435.0   
563    -122.24     37.75                45.0        891.0           435.0   
696    -122.10     37.69                41.0        746.0           435.0   

     population  households  median_income  median_house_value ocean_proximity  
290       570.0       218.0         4.3750            161900.0        NEAR BAY  
341       732.0       259.0         1.6196             85100.0        NEAR BAY  
538      3741.0      1273.0         2.5762            173400.0        NEAR BAY  
563       384.0       146.0         4.9489            247100.0        NEAR BAY  
696       387.0       161.0         3.9063            178400.0        NEAR BAY

特征total_bedrooms的缺失值用中位数435.0替换了。

5.Scikit-Learn中有SimpleImputer类，可以处理缺失值。

In [8]:
try:
    from sklearn.impute import SimpleImputer # Scikit-Learn 0.20+
except ImportError:
    from sklearn.preprocessing import Imputer as SimpleImputer

imputer = SimpleImputer(strategy="median")

特征ocean_proximity不是数值型，先去除：

In [9]:
housing_num = housing.drop('ocean_proximity', axis=1)

6.用fit方法来计算所有特征的中位数，结果保存在了变量statistics_中。

In [10]:
imputer.fit(housing_num)

SimpleImputer(add_indicator=False, copy=True, fill_value=None,
              missing_values=nan, strategy='median', verbose=0)

In [11]:
imputer.statistics_

array([-1.1849e+02,  3.4260e+01,  2.9000e+01,  2.1270e+03,  4.3500e+02,
        1.1660e+03,  4.0900e+02,  3.5348e+00,  1.7970e+05])

7.用transform方法将中位数填充到原数据中。

In [14]:
X = imputer.transform(housing_num)
housing_tr = pd.DataFrame(X, columns=housing_num.columns,
                          index = list(housing_num.index.values))
housing_tr.loc[sample_incomplete_rows.index.values]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
290    -122.16     37.77                47.0       1256.0           435.0   
341    -122.17     37.75                38.0        992.0           435.0   
538    -122.28     37.78                29.0       5154.0           435.0   
563    -122.24     37.75                45.0        891.0           435.0   
696    -122.10     37.69                41.0        746.0           435.0   

     population  households  median_income  median_house_value  
290       570.0       218.0         4.3750            161900.0  
341       732.0       259.0         1.6196             85100.0  
538      3741.0      1273.0         2.5762            173400.0  
563       384.0       146.0         4.9489            247100.0  
696       387.0       161.0         3.9063            178400.0

可以看到，缺失值已经被补足了。